util/topsql/stmtstats/stmtstats.go

```diff
@@ -0,0 +1,229 @@
+	item.KvStatsItem.KvExecCount[target] += n
+}
+
+// Take removes all existing StatementStatsMap data from StatementStats.
+// Take is thread-safe.
+func (s *StatementStats) Take() StatementStatsMap {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	data := s.data
+	s.data = StatementStatsMap{}
+	return data
+}
+
+// SetFinished marks StatementStats as "finished".
+// The background goroutine will periodically detect whether each
+// StatementStats has been finished, if so, it will be cleaned up.
+func (s *StatementStats) SetFinished() {
+	s.finished.Store(true)
+}
+
+// Finished returns whether the StatementStats has been finished.
+func (s *StatementStats) Finished() bool {
+	return s.finished.Load()
+}
+
+// BinaryDigest is converted from parser.Digest.Bytes(), and the purpose
+// is to be used as the key of the map.
+type BinaryDigest string
+
+// SQLPlanDigest is used as the key of StatementStatsMap to
+// distinguish different sql.
+type SQLPlanDigest struct {
+	SQLDigest  BinaryDigest
+	PlanDigest BinaryDigest
+}
+
+// StatementStatsMap is the local data type of StatementStats.
+type StatementStatsMap map[SQLPlanDigest]*StatementStatsItem
+
+// Merge merges other into StatementStatsMap.
+// Values with the same SQLPlanDigest will be merged.
+//
+// After executing Merge, some pointers in other may be referenced
+// by m. So after calling Merge, it is best not to continue to use
+// other unless you understand what you are doing.
+func (m StatementStatsMap) Merge(other StatementStatsMap) {
+	if m == nil || other == nil {
+		return
+	}
+	for newDigest, newItem := range other {
+		item, ok := m[newDigest]
+		if !ok {
+			m[newDigest] = newItem
+			continue
+		}
+		item.Merge(newItem)
+	}
+}
+
+// StatementStatsItem represents a set of mergeable statistics.
+// StatementStatsItem is used in a larger data structure to represent
+// the stats of a certain SQLPlanDigest under a certain timestamp.
+// If there are more indicators that need to be added in the future,
+// please add it in StatementStatsItem and implement its aggregation
+// in the Merge method.
+type StatementStatsItem struct {
+	// ExecCount represents the number of SQL executions of TiDB.
+	ExecCount uint64
+
+	// KvStatsItem contains all indicators of kv layer.
+	KvStatsItem *KvStatementStatsItem
+}
+
+// NewStatementStatsItem creates an empty StatementStatsItem.
+func NewStatementStatsItem() *StatementStatsItem {
+	return &StatementStatsItem{
+		KvStatsItem: NewKvStatementStatsItem(),
+	}
+}
+
+// Merge merges other into StatementStatsItem.
+//
+// After executing Merge, some pointers in other may be referenced
+// by i. So after calling Merge, it is best not to continue to use
+// other unless you understand what you are doing.
+//
+// If you add additional indicators, you need to add their merge code here.
+func (i *StatementStatsItem) Merge(other *StatementStatsItem) {
+	if i == nil || other == nil {
+		return
+	}
+	i.ExecCount += other.ExecCount
+	if i.KvStatsItem == nil {
```

Comment:
I guess it could be never nil as `NewKvStatementStatsItem` always allocates a map.